docs-infra: theming and dark mode

AleksanderBodurri opened the pull request:
Meant to continue the conversation from #40257.

This ~~draft~~ PR is a major refactor of the styles in https://angular.io/ in an effort to make aio themeable (primarily for a long awaited dark mode) and move towards using the sass module system.

The implementation is heavily inspired by the awesome folks who've worked on  [material.angular.io](https://github.com/angular/material.angular.io).

Previously styles were organized like this:
- styles
     - 1-layout
         - _global_layout.scss
         - ...
     - 2-modules
         - _alert.scss
         - ...

This PR changes this structure to:

- styles
     - 1-layout
         - global_layout
             - _global_layout.scss
             - _global_layout-theme.scss
         - ...
     - 2-modules
         - alert
             - _alert.scss
             - _alert-theme.scss 
         - ...

Essentially splitting styling related to theming into separate files. Within each theme file is a scss mixin that takes in a material theme configuration and generates the appropriate styling for that theme and component.

Theme configurations are defined in: `src/styles/custom-themes`

Tooling that has been ported over from [material.angular.io](https://github.com/angular/material.angular.io) allows theme chunks to be lazy loaded.

The advantage of this approach:
- Theming is contained to one file per component. Developers need only be aware of the existence of a single file when making theming related changes to a component
- Moves away from using sass `@import`s that are [no longer recommended by the sass team](https://sass-lang.com/documentation/at-rules/import)
- Themes can be lazy loaded at runtime, preventing growth of the default styles chunk every time a new theme is implemented

disadvantages:
- Splitting styles into theming files means duplication of some selectors, resulting in an increase of the default styles chunk size (in my testing this increase was about 2.6 KB on a prod build)

I'm opening this as a draft to request further technical review, and to start a conversation around what we want a potential dark mode to look like. This draft provides the tooling to make aio themeable, ~~but does not provide an actual dark mode implementation~~. I didn't want to attempt an implementation without first getting direction from the design team.

EDIT: pushed a prototype dark mode implementation to iterate from.

EDIT: PR is ready to review 🔍 

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
- [X] Feature
- [X] Code style update (formatting, local variables)
- [X] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

AleksanderBodurri committed:
feat(docs-infra): port over material io theming logic to angular io

brings in theming tools from material io into angular io in preparation of implementing darkmode

AleksanderBodurri committed:
refactor(docs-infra): rename namespaced scss to index files

scss files were forwarded from files that were named without convention, changes these file names to follow conventions

AleksanderBodurri committed:
refactor(docs-infra): grab scss constants using scss use syntax instead of the global imports

move away from using global constants in scss files because @import will be deprecated soon

AleksanderBodurri committed:
refactor(docs-infra): grab scss mixins with scss use syntax instead of global imports

move away from global mixins because @import is going to be deprecated

AleksanderBodurri committed:
feat(docs-infra): create typography and layout theme files

creates theming files for the aio typography styles; done as part of the effort to make aio themeable

AleksanderBodurri committed:
feat(docs-infra): create module theme files

creates theming files for the module styles in aio; done as part of the effor to make aio themeable

AleksanderBodurri committed:
feat(docs-infra): implement dark mode

defines styles for a first iteration of an aio darkmode

mary-poppins commented:
You can preview dd399cf at https://pr41129-dd399cf.ngbuilds.io/.
You can preview e42dbd6 at https://pr41129-e42dbd6.ngbuilds.io/.
You can preview 8cd2a5c at https://pr41129-8cd2a5c.ngbuilds.io/.

mary-poppins commented:
You can preview 77ab271 at https://pr41129-77ab271.ngbuilds.io/.

AleksanderBodurri commented:
Implemented a prototype dark mode to iterate from.

Unrelated but is the ";;" at the bottom of the page on https://angular.io/ supposed to be there? Noticed it while working on this.

<img width="1792" alt="Screen Shot 2021-03-12 at 3 08 49 PM" src="https://user-images.githubusercontent.com/39253660/110993105-dcc75d80-8344-11eb-8e19-1a362cf2f877.png">

gkalpak commented:
The `;;` at the end are definitely not supposed to be there :grin:
I accidentally introduced them in #41162 😇 They will be fixed by #41183.

mary-poppins commented:
You can preview cee8e1f at https://pr41129-cee8e1f.ngbuilds.io/.

mary-poppins commented:
You can preview 4931453 at https://pr41129-4931453.ngbuilds.io/.

mary-poppins commented:
You can preview f35ca19 at https://pr41129-f35ca19.ngbuilds.io/.

mary-poppins commented:
You can preview 85b7010 at https://pr41129-85b7010.ngbuilds.io/.

mary-poppins commented:
You can preview 93c32fc at https://pr41129-93c32fc.ngbuilds.io/.

mary-poppins commented:
You can preview 083086a at https://pr41129-083086a.ngbuilds.io/.

AleksanderBodurri commented:
Encountered a weird bug while working on this.  When I install aio with the install prompt in the url bar, the installed aio opens as expected but when I click the toggle for dark mode all the content on the screen becomes invisible. On mac os chrome 89.0.4389.90. Confirmed that this does not happen on windows.

1. Open https://pr41129-93c32fc.ngbuilds.io/
2. Install with the "install app" prompt in the url
3. Click the dark mode toggle.

Can anyone confirm being able to reproduce this on their end?

mary-poppins commented:
You can preview 7927057 at https://pr41129-7927057.ngbuilds.io/.

mary-poppins commented:
You can preview 9dcfcf5 at https://pr41129-9dcfcf5.ngbuilds.io/.

mary-poppins commented:
You can preview 8624823 at https://pr41129-8624823.ngbuilds.io/.

AleksanderBodurri commented:
> Encountered a weird bug while working on this. When I install aio with the install prompt in the url bar, the installed aio opens as expected but when I click the toggle for dark mode all the content on the screen becomes invisible. On mac os chrome 89.0.4389.90. Confirmed that this does not happen on windows.
> 
> 1. Open https://pr41129-93c32fc.ngbuilds.io/
> 2. Install with the "install app" prompt in the url bar
> 3. Click the dark mode toggle.
> 
> Can anyone confirm being able to reproduce this on their end?

Confirmed that this happens as a result of changing the background-color of the `<html>` element after the pwa opens. 
Steps to reproduce on https://angular.io/:
1) Open https://angular.io/
2) Install app with the "install app" prompt in the url bar
3) Open dev console inside pwa and manually add `background-color: #000000;` to the `<html>` element.  

I believe it's related to https://bugs.chromium.org/p/chromium/issues/detail?id=1145382

mary-poppins commented:
You can preview 9299547 at https://pr41129-9299547.ngbuilds.io/.

mary-poppins commented:
You can preview 756637a at https://pr41129-756637a.ngbuilds.io/.

mary-poppins commented:
You can preview 092483a at https://pr41129-092483a.ngbuilds.io/.

mary-poppins commented:
You can preview 6352806 at https://pr41129-6352806.ngbuilds.io/.

mary-poppins commented:
You can preview 8786d6a at https://pr41129-8786d6a.ngbuilds.io/.

mary-poppins commented:
You can preview 61052a3 at https://pr41129-61052a3.ngbuilds.io/.

twerske reviewed (commented):
This is AMAZING! Thank you for all this work @AleksanderBodurri!
I reviewed for contrast + visual design language and I have a few suggestions, but I didn't notice anything from a contrast ratio/visual a11y side - it looks like minimum 4.5 ratio is met and I ran Lighthouse/axe for contrast on a few spot check pages which all passed. 

I'll let @jelbourn weigh in on the infra of the dual theme. 

All of these are purely design suggestions or conversations (input welcome):
- We introduce a few different families of blues (warm and cool), here's a good example where I wonder if the bright blue can be toned down:
<img width="528" alt="Screen Shot 2021-03-17 at 3 44 19 PM" src="https://user-images.githubusercontent.com/15061394/111548444-189d6100-8751-11eb-9d7e-362c70581677.png">

- I'd suggest this blue on code snippets be changed to the softer blue to soften visual impact:
<img width="537" alt="Screen Shot 2021-03-17 at 3 43 06 PM" src="https://user-images.githubusercontent.com/15061394/111548449-1a672480-8751-11eb-8f1a-40ca1b3287c2.png">

- The "..." in some expansion sections borders on low contrast, can it be brightened?
<img width="247" alt="Screen Shot 2021-03-17 at 3 42 23 PM" src="https://user-images.githubusercontent.com/15061394/111548450-1affbb00-8751-11eb-80ea-3b36f5077d61.png">

- not sure about the green in the code theme personally - are we following a Material theming? Maybe we use something like the Angular Material theme from here: https://romannurik.github.io/SlidesCodeHighlighter/
<img width="553" alt="Screen Shot 2021-03-17 at 3 42 00 PM" src="https://user-images.githubusercontent.com/15061394/111548452-1b985180-8751-11eb-9ddc-0c5111b924c8.png">

- I'd suggest flipping the dark contrast here from the upper and lower portion of these knowledge boxes:
<img width="258" alt="Screen Shot 2021-03-17 at 3 41 11 PM" src="https://user-images.githubusercontent.com/15061394/111548455-1c30e800-8751-11eb-8de9-c7716d3b2872.png">

One final idea for the future is changing the Angular logo to the black variation in dark mode?

AleksanderBodurri commented:
Thanks for the review @twerske 🙏  Will find some time tomorrow to implement your suggestions

jelbourn commented:
FYI I'll try to review this sometime this week, I've got a bit of a backlog in my inbox

jelbourn commented on aio/src/app/shared/theme-picker/theme-picker.component.ts:
```diff
@@ -0,0 +1,46 @@
+// tslint:disable-next-line:max-line-length
```
The combination of theme-picker, style-manager, and theme-storage could be simplified a bit. These classes were written for mat.aio which has a bit more complex logic due to multiple themes, and it was also written before prefers-color-scheme existed. Here's my pass at creating a simplified and modernized version:
https://github.com/jelbourn/theme-picker-example/blob/master/src/app/theme-picker/theme-toggle.component.ts

Key points:
* Smaller code
* Relies on the `media` attribute of the `<link>` tag to for the initial load in order to prevent the styles jumping from one theme to another on slow network connections. 
* Moved reading `localStorage` for persisted themes to the picker's constructor since it doesn't rely on Angular's change detection at all

AleksanderBodurri replied:
Clever use of `media` to prevent render blocking from theme css. Replaced with your implementation and added the necessary `<link>` tags to `index.html`.

Is there a reason you used a custom attribute (`aio-theme`) for the initial `<link>` tags instead of a css class?

jelbourn replied:
> Is there a reason you used a custom attribute (aio-theme) for the initial <link> tags instead of a css class?
I just used an attribute out of brevity- IMO it doesn't matter a ton for attribute vs. css class here

> The first time a user visits the site they may experience a delay the first time they click the theme toggle button. 
* Interesting, I had assumed that browser caching would have prevented this, but I guess not. In that case, the picker could stop straight up deleting both `<link>` tags with the `media` attribute and instead only delete the one that doesn't matched the stored preference and remove the `media` attribute from the other (and the change the `href` on future updates).

jelbourn commented on aio/src/styles/0-base/_typography-theme.scss:
```diff
@@ -0,0 +1,56 @@
+@use '../constants' as *;
+@use '~@angular/material/theming' as *;
```
Generally you don't want to use `@use ... as *` since it's explicitly adding everything to the global namespace. It should either remain as `@import` or use `@use` with a module name. 

(here and elsewhere)

AleksanderBodurri replied:
`@import` is being deprecated in 6 months. It makes more sense to stick with `@use` to keep ahead of the deprecation.

Am I wrong in my understanding that `@use ... as *` only adds things to the top level namespace? The sass [docs](https://sass-lang.com/documentation/at-rules/use#choosing-a-namespace) point out that using `as *` for library scss is discouraged because those libraries might change and collide with variables defined in the main app. For our case `../constants` is custom scss and the material theming mixins/variables are all already prefixed with `mat`, `angular-material` or `cdk` .

The only case I can think of where mixins/functions/vars could leak implicitly is if an `@import` is used after a `@use ... as *`, which shouldn't be an issue if `@import` isn't used anymore.

In any case I'll make the change requested, I'm just curious to understand the intention behind it fully 🙏

jelbourn replied:
For this PR, we can probably keep `@import` for now and change to `@use` in a follow-up PR just for practicality. In Angular v12, we're introducing a new API for using Angular CDK/Material with `@use` that should be much nicer, so it would be nice to switch directly over to that:
https://github.com/angular/components/pull/22268

In general: you're right, it is just for the file, not global as I thought. Even then, though, I'd prefer to always namespace `@use` imports to be clear about where stuff is coming from.

jelbourn commented on aio/src/styles/0-base/_typography-theme.scss:
```diff
@@ -0,0 +1,56 @@
+@use '~@angular/material/theming' as *;
+
+@mixin typography-theme($theme) {
+    $primary: map-get($theme, primary);
```
The formatting is a bit off here; it should be +2 indent (here and in all other scss files)

jelbourn commented on aio/src/styles/1-layouts/footer/_footer-theme.scss:
```diff
@@ -0,0 +1,26 @@
+  $foreground: map-get($theme, foreground);
+  $is-dark-theme: map-get($theme, is-dark);
+
+  footer {
```
I might add a comment here explaining that the footer background is always the same color in both light and dark mode, so the foreground color doesn't need to change.

jelbourn commented on aio/src/styles/1-layouts/not-found/_not-found-theme.scss:
```diff
@@ -0,0 +1,19 @@
+@use '../../constants' as *;
+@use '~@angular/material/theming' as *;
+
+@mixin not-found-theme($theme) {
```
Some of the text on the 404 page is too low contrast

jelbourn reviewed (commented):
Some high level observations: 
* In the dark mode, the blue on the homepage and top nav doesn't feel like it quite fits with the dark background color. @twerske any thoughts on making that blue a bit darker?
* The payload size w/ this change is currently at roughly +2kb, but I'm not totally sure where that would be coming from- @gkalpak do you know if the size tracking includes the material theme? 
* I haven't worked on aio much before, but seeing all the color styles pulled out has made me realize how much ad hoc CSS there is that's not systemized in any way. It makes me think that, ideally we could reduce how much ad hoc stuff there is before pulling it out into separate theme mixins. This would manifest in having common CSS classes used across multiple components rather than each component redefining text and background colors. This could even be done in smaller, incremental PRs that could lead up to adding the theme toggle

(btw thanks for your patience, it's been a busy couple of weeks)

twerske commented:
@jelbourn @AleksanderBodurri I think making the blue on the homepage & nav darker makes sense. It's at `700` right now so I'd move it to `primary: 900` - should be a minor variable change!

jelbourn commented:
Regarding my earlier comment:
> I haven't worked on aio much before, but seeing all the color styles pulled out has made me realize how much ad hoc CSS there is that's not systemized in any way. It makes me think that, ideally we could reduce how much ad hoc stuff there is before pulling it out into separate theme mixins. This would manifest in having common CSS classes used across multiple components rather than each component redefining text and background colors. This could even be done in smaller, incremental PRs that could lead up to adding the theme toggle

I chatted with some folks on the team today about this and we decided it's okay to move ahead with this PR as-is without doing any major refactoring. aio is going to be due for a redesign at some point in the next couple years, at which point we'd need to redo all this stuff anyway.

mary-poppins commented:
You can preview 527a1b2 at https://pr41129-527a1b2.ngbuilds.io/.

AleksanderBodurri commented:
@jelbourn 

> I haven't worked on aio much before, but seeing all the color styles pulled out has made me realize how much ad hoc CSS there is that's not systemized in any way.  It makes me think that, ideally we could reduce how much ad hoc stuff there is before pulling it out into separate theme mixins. This would manifest in having common CSS classes used across multiple components rather than each component redefining text and background colors.

Totally agree. This was one of the biggest pain points in doing this refactor. Didn't feel comfortable tackling this issue because I don't have enough context on what the design of aio is **supposed** to be.

> (btw thanks for your patience, it's been a busy couple of weeks)

No worries! Thanks for the review. I've gone over some of the requested changes already, will go through the rest when I have some free time.

I addressed some of the suggestions by @twerske as well. I was only able to find 1 dark angular logo in the repo. It looks a little strange with the background. Let me know if we should replace it with something else.

<img width="1792" alt="Screen Shot 2021-04-09 at 3 58 52 AM" src="https://user-images.githubusercontent.com/39253660/114148561-e7dfc000-98e7-11eb-8816-f847284bbe0d.png">

twerske commented:
> I addressed some of the suggestions by @twerske as well. I was only able to find 1 dark angular logo in the repo. It looks a little strange with the background. Let me know if we should replace it with something else.

Agreed the logo is totally lost there - typically text is white in dark mode, could we toggle to the [inverse logo image](https://angular.io/presskit#one-color-inverse-logo)?

mary-poppins commented:
You can preview 0c65282 at https://pr41129-0c65282.ngbuilds.io/.

mary-poppins commented:
You can preview b1ea9dc at https://pr41129-b1ea9dc.ngbuilds.io/.

mary-poppins commented:
You can preview 00eec53 at https://pr41129-00eec53.ngbuilds.io/.

AleksanderBodurri commented:
@twerske Changed to the inverse logo image. Take a look at the latest aio ci build and let me know what you think (it may be a little too bright for a dark mode?). Can you also take a look at the api list page and let me know what you think of the colors? Also I'm wondering what we should do for the icons at the top right of the page when viewing on mobile. Should we hide some of them so there isn't an obvious overflow? 

@jelbourn I've addressed your comments. Let me know if theres anything else you'd like me to do for this PR.

twerske commented:
> @twerske Changed to the inverse logo image. Take a look at the latest aio ci build and let me know what you think (it may be a little too bright for a dark mode?). Can you also take a look at the api list page and let me know what you think of the colors? Also I'm wondering what we should do for the icons at the top right of the page when viewing on mobile. Should we hide some of them so there isn't an obvious overflow?

SVG looks great. For context, dark mode having accent white text/images is definitely the correct visual guideline, so it's not too bright! Can you add a background fill under the shield to the SVG of #303030 so that there's no slash through the logo? In certain responsive sizes it reads a little like we're crossing out the Angular logo. 

API colors LGTM.

Re: icon overflow.... I'd prioritize for mobile in this order: Theme switcher, Twitter, YouTube, GitHub (If the screen is small, Youtube & github may not be most useful). Maybe @jelbourn @mgechev have thoughts on the icons in small phone screens?

Two other comments observed:
1. Looks like the icons are switched, should be a sun for switching to light mode and a moon for switching to dark? 
2. I tested with my system preferences and it looks like we're not incorporating system preference into defaulting to dark mode?

mary-poppins commented:
You can preview 807be45 at https://pr41129-807be45.ngbuilds.io/.

AleksanderBodurri commented:
@twerske Changed to a logo with a solid fill for dark mode. Addressed the other requests as well. Take a look at the latest CI build and let me know what you think. 🙏 

@jelbourn FYI I had to change the `bundleName` for the styles array in `angular.json` by removing the `assets/` prefix because of this issue https://github.com/angular/angular-cli/issues/20360.

jelbourn reviewed (commented):
Overall looks great, just a few last minor comments from me

For the icons in the top bar, is it me or are the icons slightly off?
![image](https://user-images.githubusercontent.com/838736/115640278-6f0b3b80-a2cb-11eb-803e-9245f79dd02f.png)

I can't tell if it's the vertical centering or if it's that the icons are different sizes

mary-poppins commented:
You can preview 0c86a03 at https://pr41129-0c86a03.ngbuilds.io/.

mary-poppins commented:
You can preview 4e51b11 at https://pr41129-4e51b11.ngbuilds.io/.

AleksanderBodurri commented:
@jelbourn 
> For the icons in the top bar, is it me or are the icons slightly off?
>I can't tell if it's the vertical centering or if it's that the icons are different sizes

I think it's because they're different sizes.

I've addressed your other comments. Is it unusual that the `main-es2015` payload went from 451829 bytes to 509750? This seems like a drastic increase.

jelbourn commented:
Yeah, that size increase is quite suspicious- @gkalpak do you know how the aio size here is computed? Any advice for looking for the source of increase?

petebacondarwin commented:
> Yeah, that size increase is quite suspicious- @gkalpak do you know how the aio size here is computed? Any advice for looking for the source of increase?

The payload is literally the uncompressed `main.js` that is created by the CLI when bundling. I usually make a build of master and a build of the PR via:

```
cd aio
yarn build-local
```

And then try comparing the two files.

AleksanderBodurri commented:
@jelbourn Looks like `MatTooltipModule` is the cause of the size bump. See build and sourcemap comparisons below. How do you want to proceed?

Build without tooltip module
<img width="997" alt="Screen Shot 2021-04-30 at 9 27 34 PM" src="https://user-images.githubusercontent.com/39253660/116766919-0aeb2480-a9fb-11eb-8d66-4e77cbbc505c.png">

Build with tooltip module (approximately 55.6kb increase for `main-es2015`)
<img width="1002" alt="Screen Shot 2021-04-30 at 9 27 44 PM" src="https://user-images.githubusercontent.com/39253660/116766907-fc047200-a9fa-11eb-86fd-15a507cc762a.png">

Full source map comparison (without tooltip module on the left / with tooltip module on the right)
<img width="1792" alt="Screen Shot 2021-04-30 at 9 46 49 PM" src="https://user-images.githubusercontent.com/39253660/116767342-95348800-a9fd-11eb-8b07-b583fda795d5.png">

Angular source map comparison (without tooltip module on the left / with tooltip module on the right)
<img width="1792" alt="Screen Shot 2021-04-30 at 9 39 13 PM" src="https://user-images.githubusercontent.com/39253660/116767223-eb54fb80-a9fc-11eb-9ef9-5831793ee40a.png">

Cdk source map comparison (without tooltip module on the left / with tooltip module on the right)
<img width="1791" alt="Screen Shot 2021-04-30 at 9 41 50 PM" src="https://user-images.githubusercontent.com/39253660/116767257-1dfef400-a9fd-11eb-953b-56a62b5237f8.png">

Material source map comparison (without tooltip module on the left / with tooltip module on the right)
<img width="1792" alt="Screen Shot 2021-04-30 at 9 40 49 PM" src="https://user-images.githubusercontent.com/39253660/116767272-26efc580-a9fd-11eb-9e24-610d3cf4eb68.png">

AleksanderBodurri commented:
@jelbourn One thing I forgot to mention,

I took a css-only approach to change the Angular image based on the current theme. I changed the `<img>` to a `<div>` with a css background image applied based on the theme css that was loaded in. The downside to this is that we can't use an `alt` attribute. We could use an `aria-label` and `role="img"` on the div instead, if we proceed with this solution. The current solution also conflicts with a change in PR https://github.com/angular/angular/pull/34900  that adds the `loading="lazy"` attribute to the logo. 

A programatic approach could be done instead. We could change the `src` of the `<img>` when the theme changes. I didn't go this route initially because It felt more natural to only make changes for a theme in the css file associated with that theme. It's a little messy because we'd have to query the DOM for the logo (which won't exist on some pages) from the theme toggle component. Still the benefit of having the logo as an `<img>` might be worth a little more complexity.

Which solution do you think we should go with?

jelbourn commented:
For the tooltip, that makes sense- if nothing else on aio is using cdk/overlay, it would pull in a lot of stuff. I would probably remove `matTooltip` and replace it with the native `title` element just to save those bytes. 

For the image, IMO it's a purely decorative image and isn't _content_, so I think think it's a problem if it's just a `div` with a background-image. The `lazy` attribute also isn't going to really go anything since the image is literally the first thing on the page.

AleksanderBodurri commented:
> For the tooltip, that makes sense- if nothing else on aio is using cdk/overlay, it would pull in a lot of stuff. I would probably remove `matTooltip` and replace it with the native `title` element just to save those bytes.

Sounds good.

> For the image, IMO it's a purely decorative image and isn't _content_, so I think think it's a problem if it's just a `div` with a background-image. 

Do you mean to say that you **don't** think its a problem if its just a div? I agree that it's decorative and WCAG guidelines say the css background image approach is preferred in that case. Just want to make sure before I proceed. I assume this would also mean to not add the aria-label or `role="img"` attributes.

jelbourn commented:
Oh yeah, that was an unfortunate word to typo. I think using a div w/ `background-image` is okay in this particular case.

mary-poppins commented:
You can preview cc8ef67 at https://pr41129-cc8ef67.ngbuilds.io/.

mary-poppins commented:
You can preview aab8e7a at https://pr41129-aab8e7a.ngbuilds.io/.

mary-poppins commented:
You can preview dde3838 at https://pr41129-dde3838.ngbuilds.io/.

AleksanderBodurri commented:
@jelbourn not sure why but it looks like `media` for theme link elements is being transformed during the build process
```html
  <link aio-theme="light" href="light-theme.css" rel="stylesheet" media="(prefers-color-scheme: light)">
```

becomes

```html
<link aio-theme="light" href="light-theme.css" rel="stylesheet" media="print" onload="this.media='(prefers-color-scheme: light)'"><noscript><link rel="stylesheet" href="light-theme.css" media="(prefers-color-scheme: light)"></noscript>
```

The onload event firing is an issue because it's not possible to know exactly when the media attribute of the link will set to `(prefers-color-scheme: ...)`. Because of this, removing the media attribute like you suggested above is unreliable (we might remove the media attribute before the onload event reapplies it). I'm not sure why this transformation is occurring, I don't think it was happening in previous builds. Any advice?

mary-poppins commented:
You can preview 1646156 at https://pr41129-1646156.ngbuilds.io/.

alan-agius4 commented on aio/src/app/shared/theme-picker/theme-toggle.component.ts:
```diff
@@ -0,0 +1,91 @@
+    if (storedPreference) {
+      this.isDark = storedPreference === 'true';
+    } else {
+      this.isDark = matchMedia?.('(prefers-color-scheme: dark)').matches ?? false;
```
Should we add an event listener to sync with theme with the System? In most cases, the theme changes based on the time of the day.

Using local storage in this case will also cause the theme not to be synced with the system on refresh.

I don't know if we want to go down this route but IMHO theming has 3 states and not 2.

- Auto
- Dark
- Light

jelbourn replied:
> In most cases, the theme changes based on the time of the day

What is this based on? While Apple and Android devices support this option, but I think the default is "light" and people can choose either "dark" or "auto" (manually setting up a schedule on Android). On Windows "auto" is not an option. ChromeOS doesn't support system themes yet. 

> Using local storage in this case will also cause the theme not to be synced with the system on refresh.

That was the intent- if a user never touches the theme toggle, it will respect their system setting. If they manually change the theme, it will remember their choice. I suppose we could drop the part about remembering the user's choice, but my experience has always been that I want sites to retain this preference (especially since not all OSes support dark mode yet). 

IMO adding extra code to react to changes in the system theme is overkill. 

@AleksanderBodurri the media issue there is due to the critical CSS inlining built into Angular CLI. I see two options:
* Turn off that feature
* Change from using the `media` attribute to having a single `link` element that does something like this:
```scss
@import url("light-theme.css") (prefers-color-scheme: light);
@import url("dark-theme.css") (prefers-color-scheme: dark);
```

Both of these approaches effectively do the same thing- force the CSS into a separate load. I don't see any way around the inherent limitation of media queries being a purely client-side construct here.

mary-poppins commented:
You can preview 9735e6b at https://pr41129-9735e6b.ngbuilds.io/.

mary-poppins commented:
You can preview bfd19f9 at https://pr41129-bfd19f9.ngbuilds.io/.

AleksanderBodurri commented:
Updated to work around the critical css inlining feature for the theming css. For some reason it looks like the critical css being inlined from `styles.scss` is not being parsed correctly. You can notice this on page load where some of the elements are viewable and unstyled for a very brief (but noticable) moment until the full `styles.css` file loads in. I'm not sure if that's an issue with this PR specifically. If it is, I'm unsure what could be causing it. I'm not familiar with how critical css inlining is implemented. Maybe @alan-agius4 could help? 🙏 

![demo](https://user-images.githubusercontent.com/39253660/117586256-0ef70080-b0e5-11eb-83a2-56c64f8dfbd0.png)

jelbourn commented on aio/src/app/shared/theme-picker/theme-toggle.component.ts:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    const initialTheme = document.querySelector('#aio-initial-theme');
+    initialTheme?.remove?.();
```
`remove` is unfortunately not available in IE11, which we're supporting until v13

AleksanderBodurri replied:
good catch

jelbourn reviewed (approved):
LGTM, using the `<style>` element with `@import` is even a bit cleaner than have the two link elements since you only have to remove the one element. 

@gkalpak and @aikidave it would be great to get your review on this one

mary-poppins commented:
You can preview 9ad32e2 at https://pr41129-9ad32e2.ngbuilds.io/.

mary-poppins commented:
You can preview ba11f12 at https://pr41129-ba11f12.ngbuilds.io/.

crisbeto commented:
Just an FYI: there may be some incoming merge conflicts after we've run `ng update` on AIO to switch it to v12 final. Angular Material has an automated migration to switch to the new theming Sass API.

twerske reviewed (approved):
LGTM from the devrel/a11y perspective

aikithoughts reviewed (approved):
Reviewed-for: global-docs-approvers

jessicajaniuk reviewed (approved):
LGTM 🍪 

reviewed-for: size-tracking

petebacondarwin reviewed (approved):
There are now conflicts to resolve, but otherwise LGTM

Reviewed-for: size-tracking

mary-poppins commented:
You can preview d1a55f5 at https://pr41129-d1a55f5.ngbuilds.io/.

jelbourn commented:
@gkalpak any final comments on this?

gkalpak commented on aio/src/styles/1-layouts/marketing-layout/_marketing-layout-theme.scss:
```diff
@@ -0,0 +1,73 @@
+    color: constants.$white;
+
+    .hero-logo {
+      background-image: if($is-dark-theme, url(/assets/images/logos/angular/angular_solidWhite.svg),url(/assets/images/logos/angular/angular.svg));
```
FWIW, I don't think using a white-background image is a good idea. Angular's red shield is very characteristic and I think  we should keep it.

What was the reasoning for using a different logo on dark mode. AFAICT, the original logo looks fine (but I might be missing something): ![original-logo-on-dark](https://user-images.githubusercontent.com/8604205/118854146-34ea8500-b8dd-11eb-90db-b123e0df3b21.png)

gkalpak commented on aio/src/app/app.component.html:
```diff
@@ -25,6 +24,7 @@
     <aio-top-menu *ngIf="showTopMenu" [nodes]="topMenuNodes" [currentNode]="currentNodes?.TopBar"></aio-top-menu>
     <aio-search-box class="search-container" #searchBox (onSearch)="doSearch($event)" (onFocus)="doSearch($event)"></aio-search-box>
     <div class="toolbar-external-icons-container">
+      <aio-theme-picker></aio-theme-picker>
```
This placement ends up being a little confusing imo, because the theme-picker looks like another external/social icon. (It is also strange to put this inside a container labeled as "toolbar-**external**-icons-container" :grin:

I would move the theme-picker in a separator container and ensure there is some visual separator between it and the social icons (via a right border or something).

gkalpak reviewed (approved):
Amazing work, @AleksanderBodurri (and everyone chiming in/reviewing) 🤯 :100: 🥇 

I have left a few comments, but I would be fine with dealing with (most of) them in a follow-up PR (since this has already gotten quite big).

Also, I noticed a couple of issues on smaller screens (due to increasing the items we try to cram in the limited horizontal space of the toolbar.

##
At around 480px width, the search-bar overlaps with the Angular logo:

![searchbar-issue-on-small](https://user-images.githubusercontent.com/8604205/118854143-3451ee80-b8dd-11eb-9909-14f51bdc54b8.png)

##
On mobile phones (e.g. at 360px width), some of the social icons overflow out of the screen (with no indication they exist or a way to get to them). Maybe we could consider moving them into the sidenav on smaller screens?

![homescreen-on-mobile](https://user-images.githubusercontent.com/8604205/118854139-3320c180-b8dd-11eb-8d4b-6f5fbeb65acd.png)

AleksanderBodurri commented on aio/src/styles/2-modules/toc/_toc.scss:
```diff
@@ -1,12 +1,13 @@
 $tocItemTopPadding: 9;
-$tocMarkerRailSize: 1;
 $tocMarkerSize: 6;
+$tocMarkerRailSize: 1;
```
Is there somewhere else it should be?

AleksanderBodurri replied:
Oh nvm. I thought you meant a different file. The line moved during the refactor, should be fixed now.

AleksanderBodurri commented on aio/src/app/shared/theme-picker/theme-toggle.component.ts:
```diff
@@ -0,0 +1,91 @@
+}
+
+@Component({
+  selector: 'aio-theme-picker',
```
Good catch

AleksanderBodurri commented on aio/src/app/shared/theme-picker/theme-toggle.component.ts:
```diff
@@ -0,0 +1,91 @@
+      this.isDark = storedPreference === 'true';
+    } else {
+      this.isDark = matchMedia?.('(prefers-color-scheme: dark)').matches ?? false;
+      this.themeStorage.setThemePreference(this.isDark);
```
Good catch

mary-poppins commented:
You can preview 9d79153 at https://pr41129-9d79153.ngbuilds.io/.

AleksanderBodurri commented:
@gkalpak I think I've addressed most of your comments. There was one I was uncertain of so I left it open for clarification. 

Appreciate the review 🙏 Let me know if theres anything else I can do for this PR.

mary-poppins commented:
You can preview 9a73356 at https://pr41129-9a73356.ngbuilds.io/.

mary-poppins commented:
You can preview c800bca at https://pr41129-c800bca.ngbuilds.io/.

mary-poppins commented:
You can preview c918995 at https://pr41129-c918995.ngbuilds.io/.

mary-poppins commented:
You can preview 7ca90ae at https://pr41129-7ca90ae.ngbuilds.io/.

gkalpak reviewed (approved):
Thank you for addressing all the comments, @AleksanderBodurri ❤️
I don't see anything that should block this PR. Incredible work 💯

I can't tell you how excited I am for this feature 🤤 🚀 🕙 

:shipit:
